Add a new diagnostic variable dvsdtd.

dabail10 opened the pull request:
For detailed information about submitting Pull Requests (PRs) to the CICE-Consortium,
please refer to: <https://github.com/CICE-Consortium/About-Us/wiki/Resource-Index#information-for-developers>


## PR checklist
- [X] Short (1 sentence) summary of your PR: 
This adds a new diagnostic variable dvsdtd so that the SIMIP variable sndmassdyn can be used.
- [X] Developer(s): 
dabail10 (D. Bailey)
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#1db0d9c13c0885c2da5c498a3fee0daa60788434
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [X] No
- Does this PR update the Icepack submodule?  If so, the Icepack submodule must point to a hash on Icepack's main branch.
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [ ] Yes
    - [X] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [X] No 
- [X] Please document the changes in detail, including _why_ the changes are made.  This will become part of the PR commit log.
We finally added the SIMIP variable sndmassdyn. This required the computation of dvsdtd. Also, there were some unit issues with some of the sndmass variables.

eclare108213 commented on cicecore/cicedyn/general/ice_flux.F90:
```diff
@@ -419,6 +420,7 @@ subroutine alloc_flux
          strintyU   (nx_block,ny_block,max_blocks), & ! divergence of internal ice stress, y (N/m^2)
          daidtd     (nx_block,ny_block,max_blocks), & ! ice area tendency due to transport   (1/s)
          dvidtd     (nx_block,ny_block,max_blocks), & ! ice volume tendency due to transport (m/s)
+         dvsdtd     (nx_block,ny_block,max_blocks), & ! snow volume tendency due to transport (m/s)
```
When I first looked at the PR, I thought this was salt, not snow.  Do we need to distinguish the names of those now, in case we want to add a salt tendency term?

dabail10 replied:
I guess it would not be hard to add a dvsdtt term. I was doing it specifically since there is the need for an SIMIP variable sndmassdyn. There is not a similar term for the thermodynamics. The snow melt and snow ice formation are called out in separate terms. I was using dvsdtd to be consistent with dvidtd.

eclare108213 reviewed (commented):
Looks good.  Is there not also a dvsdtt?
These changes will need to be made to all of the drivers, eventually.

apcraig reviewed (commented):
In update state, I suggest
- Add "if present" to dagedt calculation
- With "if present", I think this syntax still sometimes creates run time problems with debug flags (it shouldn't)

```
   if (present(xxx)) xxx = yyy
```
   would recommend


```
   if (present(xxx)) then
     xxx = yyy
   endif
```
- Can we update the indentation in update_state?  Seems it may be overdue.

apcraig commented:
> Looks good. Is there not also a dvsdtt? These changes will need to be made to all of the drivers, eventually.

Since update_state is not backwards compatible, could you update all the call to update_state in all the other drivers.  I would NOT add dvsdtd to those calls, but just update the arguments to use keywords= syntax.  That should be adequate.  Thanks.

dabail10 commented:
Ah. Good point. I will do that. I will also work on dvsdtt at the same time.

dabail10 commented:
I have added dvsdtt and fixed all of the calls to update_state in the drivers. The updated test results for the standalone driver are here:

https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#e68dd144aa1577111f5a87f93378afd15847b62e

apcraig commented:
I ran a test suite on Derecho and this all looks good.  @dabail10, anything left to do?  Any other comments?  If not, I'll merge in the next day or two.

dabail10 commented:
I believe this is everything here.
